doc/guide/transactions-and-threading.rst

```diff
@@ -0,0 +1,394 @@
+
+.. -> src
+
+   >>> exec(src)
+   >>> conn.root.x
+   1
+   >>> conn.close()
+
+In this example, because we aborted the transaction, the value of
+``conn.root.x`` was rolled back to 1.
+
+There are a number of things going on here that deserve some
+explanation.  When using transactions, there are three kinds of
+objects involved:
+
+Transaction
+   Transactions represent units of work.  They have beginnings and
+   ends. They provide the
+   :interface:`~transaction.interfaces.ITransaction` interface.
+
+Transaction manager
+   Transaction managers create transactions and
+   provide APIs to start and end transactions.  The transactions
+   managed are always sequential. There is always exactly one active
+   transaction associated with a transaction manager at any point in
+   time. Transaction managers provide the
+   :interface:`~transaction.interfaces.ITransactionManager` interface.
+
+Data manager
+   Data managers manage data associated with transactions.  ZODB
+   connections are data managers.  The details of how they interact
+   with transactions isn't important here.
+
+Explicit transaction managers
+-----------------------------
+
+ZODB connections have transaction managers associated with them when
+they're opened. When we call the database :meth:`~ZODB.DB.open` method
+without an argument, a thread-local transaction manager is used. Each
+thread has it's own transaction manager.  When we called
+``transaction.commit()`` above we were calling commit on the
+thread-local transaction manager.
+
+Because we used a thread-local transaction manager, all of the work in
+the transaction needs to happen in the same thread.  Similarly, only
+one transaction can be active in a thread.
+
+If we want to have transactions who's work is spread over multiple
+threads, or if we wanted to run multiple simultaneous transactions in
+a single thread, then we can create transaction managers ourselves and
+pass them to :meth:`~ZODB.DB.open`::
+
+  tm = transaction.TransactionManager()
+  conn = db.open(tm)
+  conn.root.x = 2
+  tm.commit()
+
+.. -> src
+
+   >>> exec(src)
+
+In this example, to commit our work, we called ``commit()`` on the
+transaction manager we created and passed to :meth:`~ZODB.DB.open`.
+
+context managers
+----------------
+
+In the examples above, the transaction beginnings were
+implicit. Transaction's were effectively
+[#implicit-transaction-creation]_ created when the transaction
+managers were created and when previous transactions were committed.
+We can create transactions explicitly using
+:meth:`~transaction.interfaces.ITransactionManager.begin`::
+
+  tm.begin()
+
+.. -> src
+
+   >>> exec(src)
+
+A more modern [#context-managers-are-new]_ way to manage transaction
+boundaries is to use context managers and the Python ``with``
+statement. Transaction managers are context managers, so we can use
+them with the ``with`` statement directly::
+
+  with tm as trans:
+     trans.note("incrementing x")
+     conn.root.x += 1
```

Comment:
I ended up keeping the description of using transaction managers as context managers, because it made it a bit easier to provide a retry example.  I also used a longer variable name for the transaction manager, which hopefully helps.